### ISSUE TITLE
Gathering useful heap profiles MIPS/Linux/uclibc

### ISSUE THREAD

**Issue by zanderso** [gperftools/gperftools#822]
We're trying to gather heap profiles of the Dart VM on some MIPS hardware running Linux and uclibc to try and help debug an unexplained growth in the size of the C heap. We have configured tcmalloc with `--enable-emergency-malloc`, and we're running with `HEAP_PROFILE_TIME_INTERVAL=60`. Heap profiles are emitted, but they don't contain much information, e.g. after the first minute:

```
heap profile:  18949: 20815390 [177357: 799048871] @ heapprofile
18949: 20815390 [177357: 799048871] @

MAPPED_LIBRARIES:
00400000-00fe3000 r-xp 00000000 00:00 4873        /usr/bin/dart
00ff3000-01026000 r--p 00be3000 00:00 4873        /usr/bin/dart
01026000-01069000 rw-p 00c16000 00:00 4873        /usr/bin/dart
01069000-0106c000 rwxp 00000000 00:00 0           [heap]
```

In this example the Dart VM is unstripped and contains debugging info. The VM does lots of calls to malloc and new during initialization, so we were expecting to see a bit more detail.

**Comment by zanderso** [comment]
@alk asked:

> I see two possible reasons of empty heap profile:
> a) dart isn't using malloc much (after all it's heavily gc-ful runtime, right?)
> b) 60 seconds isn't enough to overflow it's buckets and trigger dumping stuff to disk (normal exit(0) of the process does flush everything).

**Comment by zanderso** [comment]
As for (a): The Dart heap where we allocate Dart objects we mmap and manage ourselves. Our own introspection tools indicate that this is not where the leak is. We use the C heap for internal data structures, e.g. IO buffers, data structures used during compilation, and other bookkeeping information. We also call out to various libraries such as boringssl and zlib, which probably do some allocation in the C heap. We've inspected the process memory map after the unexplained memory growth and found that the growth is in the C heap.

As for (b): Could you elaborate on this a bit. I believe John has set an environment variable to dump a profile every 60 seconds (HEAP_PROFILE_TIME_INTERVAL). This is the profile he pasted above. Do we need to set another parameter for the profiles dumped every 60 seconds to contain more information?

**Comment by zanderso** [comment]
From John, it looks like the backtraces are working:

```
GTAFNS1441P0196:/rw# ./stacktrace_unittest
Obtained 8 stack frames.
Obtained 8 stack frames.
/user/rw/stacktrace_unittest [0x40117c] 0x40117c
/user/rw/stacktrace_unittest [0x4014ac] 0x4014ac
/user/rw/stacktrace_unittest [0x4015dc] 0x4015dc
/user/rw/stacktrace_unittest [0x40170c] 0x40170c
/user/rw/stacktrace_unittest [0x40183c] 0x40183c
/user/rw/stacktrace_unittest [0x40194c] 0x40194c
/user/rw/stacktrace_unittest(main+0x1c) [0x400ce4] 0x400ce4
/lib/libc.so.1(__uClibc_main+0x204) [0x76e341c4] 0x76e341c4
CheckStackTrace() addr: 0x4018f0
Backtrace 0: expected: 0x401164..0x4012c8  actual: 0x40117c ... OK
Backtrace 1: expected: 0x40149c..0x4014b4  actual: 0x4014ac ... OK
Backtrace 2: expected: 0x4015c8..0x4015e4  actual: 0x4015dc ... OK
Backtrace 3: expected: 0x4016f8..0x401714  actual: 0x40170c ... OK
Backtrace 4: expected: 0x401828..0x401844  actual: 0x40183c ... OK
Backtrace 5: expected: 0x401934..0x401954  actual: 0x40194c ... OK
PASS
```

**Comment by alk** [comment]
Yes, that means stacktraces are not captured by heap profiler. Yet, stacktrace_unittest is working.

Is it possible that some of the code is built without unwinding info ? (e.g. maybe something like -fno-exceptions) I am not mips expert but if unwind info is not on by default, it could be disabled by -fno-exceptions.

**Comment by zanderso** [comment]
Ah, okay. Yah, we build the VM with -fno-exceptions:

https://github.com/dart-lang/sdk/blob/master/tools/gyp/configurations_make.gypi#L25

John, can you try again without that flag?

@alk Do you see any other flags there that would be problematic?

**Comment by alk** [comment]
You may want to try explicit -fexceptions. Or -funwind-tables looks like even better option.

**Comment by jtmcdole** [comment]
@zanderso I'll rebuild this morning.

**Comment by jtmcdole** [comment]
> b)  60 seconds isn't enough to overflow it's buckets and trigger dumping stuff to disk (normal exit(0) of the process does flush everything).

So the timeout interval wont trigger this at all? Is there a signal we can send the process to get the profile?

**Comment by alk** [comment]
I was actually wrong about overflows. If you do  HEAP_PROFILE_TIME_INTERVAL and your program uses malloc then you will get periodic heap dumps. And from description above it does look like you're having them. So clearly this isn't "stuff isn't flushed" problem.

**Comment by jtmcdole** [comment]
Looks like we're still not getting stacks, even with -funwind-tables  Will check with Zach

```
heap profile:   1191: 10659726 [170451: 746708436] @ heapprofile
  1191: 10659726 [170451: 746708436] @

MAPPED_LIBRARIES:
00400000-010f7000 r-xp 00000000 00:00 3346        /user/rw/dart
01106000-01157000 r--p 00cf6000 00:00 3346        /user/rw/dart
01157000-0119a000 rw-p 00d47000 00:00 3346        /user/rw/dart
0119a000-0119d000 rwxp 00000000 00:00 0           [heap]
71dec000-71ff6000 rw-p 00000000 00:00 0           
72378000-72379000 ---p 00000000 00:00 0           
72379000-72578000 rw-p 00000000 00:00 0           [stack:32439]
72638000-72a58000 rw-p 00000000 00:00 0           
72ed8000-73108000 rw-p 00000000 00:00 0           
73109000-73119000 rw-p 00000000 00:00 0           
73219000-73229000 rw-p 00000000 00:00 0           
7323a000-7335a000 rw-p 00000000 00:00 0           
```

**Comment by zanderso** [comment]
We set `PERFTOOLS_VERBOSE=10` and were expecting to see some log messages like, "Found hooked allocator at..." from `MallocHook_GetCallerStackTrace()` but did not see any. We also did not see any warnings like "Hooked allocator frame not found, returning empty trace". I guess this means that `GetStackTrace()` is just always failing and returning 0?

**Comment by alk** [comment]
This is quite interesting. So we have stacktrace_unittest with working stacktraces and dart vm without. Apparently. I am not at all familiar with mips. There has to be some compilation or linking flag or something else (size of binary? luck?) that causes this.

Looking at your "mapped libraries" above I am wondering if your guys are linking dart fully statically and if this is contributing in some way. It might be worth trying to play with building and linking stacktrace unittest with various flags and/or fully statically.

Do you guys have e.g. qemu image or something along this lines that I could try playing with myself ?

**Comment by zanderso** [comment]
I was able to play around with the flags a bit, and seem to have useful backtraces now. The solution is to pass '-fexceptions' and '-funwind-tables' when building tcmalloc as well as the Dart VM. Without this, the backtrace would make it up only the first few frames. I guess libgcc backtracce on MIPS can't make it across the boundary from the .so to the caller without unwind info on both sides. Here is a summary of the ./configure command line I used for tcmalloc:

```
./configure --enable-emergency-malloc CFLAGS="-fexceptions -funwind-tables"
```

The only change on the Dart VM side was to add "-fexceptions and -funwind-tables" and remove "-fno-exceptions".
